### ISSUE TITLE
[BUG] MapHttpAdapter is always null so no data from HttpContext is logged

### ISSUE THREAD

**Issue by olavaukan** [elastic/ecs-dotnet#62]
**ECS version**: 1.4.0

**Elasticsearch version (if applicable)**: 7.6.0

**.NET framework / OS**: .NET Core 3.1 / Windows 10 and CentOS (Docker)

**Description of the problem, including expected versus actual behavior**:
When used in a ASP.NET Core 3.1 MVC/API-project the Elastic.CommonSchema.Serilog.EcsTextFormatter is unable to pickup any info from HttpContext so this data is not added to the JSON-serialized log output. Debugging LogEventConverter indicates that configuration.MapHttpAdapter is null, so it does not try to get any data from HttpContext.

**Steps to reproduce**:
1. Using Visual Studio 2019 -> File -> New Project -> ASP.NET Core Web Application.
2. Select API template.
3. Dependencises -> Manage NuGet Packages -> Add reference to NuGet packages Serilog.AspNetCore, Serilog.Sinks.Console, Serilog.Settings.Configuration and Elastic.CommonSchema.Serilog 
4. Change Startup.cs and Program.cs according to Serilog docs https://github.com/serilog/serilog/wiki/Getting-Started
5. Configure EcsTextFormatter in Program.cs  as described in ecs-dotnet docs https://github.com/elastic/ecs-dotnet/tree/master/src/Elastic.CommonSchema.Serilog 
6. Run/debug solution using Kestrel (Select WebApplication1, not IIS Express).
7. Confirm that the console output does not contain any of the HttpContext data such as Server, url, UserAgent etc.

Program.cs:
```
using System;
using System.IO;
using Elastic.CommonSchema.Serilog;
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Serilog;

namespace WebApplication1
{
	public class Program
	{
		public static IConfiguration Configuration { get; } = new ConfigurationBuilder()
			.SetBasePath(Directory.GetCurrentDirectory())
			.AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
			.AddEnvironmentVariables()
			.Build(); 

		public static IWebHost BuildWebHost(string[] args) =>
			WebHost.CreateDefaultBuilder(args)
				.UseStartup<Startup>()
				.UseSerilog((ctx, config) =>
				{
					config.ReadFrom.Configuration(ctx.Configuration);
					config.WriteTo.Console(new EcsTextFormatter());
				})
				.UseKestrel()
				.Build();

		public static void Main(string[] args)
		{
			Log.Logger = new LoggerConfiguration()
				.ReadFrom.Configuration(Configuration)
				.Enrich.WithProperty("App Name", "Serilog Web App Sample")
				.CreateLogger();
			try
			{
				BuildWebHost(args).Run();
				return;
			}
			catch (Exception ex)
			{
				Log.Fatal(ex, "Host terminated unexpectedly");
				return;
			}
			finally
			{
				Log.CloseAndFlush();
			}
		}
	}
}

```
Startup.cs:
```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace WebApplication1
{
	public class Startup
	{
		public Startup(IConfiguration configuration)
		{
			Configuration = configuration;
		}

		public IConfiguration Configuration { get; }

		// This method gets called by the runtime. Use this method to add services to the container.
		public void ConfigureServices(IServiceCollection services)
		{
			services.AddControllers();
		}

		// This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
		public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
		{
			if (env.IsDevelopment())
			{
				app.UseDeveloperExceptionPage();
			}

			app.UseRouting();

			app.UseAuthorization();

			app.UseEndpoints(endpoints =>
			{
				endpoints.MapControllers();
			});
		}
	}
}

```
Console output:
```
{
	"@timestamp": "2020-03-11T12:31:42.7001953+01:00",
	"log.level": "Information",
	"message": "Request starting HTTP/1.1 GET http://localhost:5000/weatherforecast  ",
	"_metadata": {
		"message_template": "{HostingRequestStartingLog:l}",
		"protocol": "HTTP/1.1",
		"method": "GET",
		"content_type": null,
		"content_length": null,
		"scheme": "http",
		"path_base": "",
		"path": "/weatherforecast",
		"query_string": "",
		"hosting_request_starting_log": "Request starting HTTP/1.1 GET http://localhost:5000/weatherforecast  ",
		"event_id": {
			"id": 1
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7001953+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Hosting.Diagnostics",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7114543+01:00",
	"log.level": "Information",
	"message": "Executing endpoint '\"WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)\"'",
	"_metadata": {
		"message_template": "Executing endpoint '{EndpointName}'",
		"endpoint_name": "WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)",
		"event_id": {
			"name": "ExecutingEndpoint"
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7114543+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Routing.EndpointMiddleware",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7244166+01:00",
	"log.level": "Information",
	"message": "Route matched with \"{action = \\\"Get\\\", controller = \\\"WeatherForecast\\\"}\". Executing controller action with signature \"System.Collections.Generic.IEnumerable`1[WebApplication3.WeatherForecast] Get()\" on controller \"WebApplication3.Controllers.WeatherForecastController\" (\"WebApplication3\").",
	"_metadata": {
		"message_template": "Route matched with {RouteData}. Executing controller action with signature {MethodInfo} on controller {Controller} ({AssemblyName}).",
		"route_data": "{action = \"Get\", controller = \"WeatherForecast\"}",
		"method_info": "System.Collections.Generic.IEnumerable`1[WebApplication3.WeatherForecast] Get()",
		"controller": "WebApplication3.Controllers.WeatherForecastController",
		"assembly_name": "WebApplication3",
		"event_id": {
			"id": 3,
			"name": "ControllerActionExecuting"
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"id": "d756316f-e9e8-45f3-bbea-9b2fe6986f5a",
		"action": "WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)",
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7244166+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7361828+01:00",
	"log.level": "Information",
	"message": "Executing ObjectResult, writing value of type '\"WebApplication3.WeatherForecast[]\"'.",
	"_metadata": {
		"message_template": "Executing ObjectResult, writing value of type '{Type}'.",
		"type": "WebApplication3.WeatherForecast[]",
		"event_id": {
			"id": 1,
			"name": "ObjectResultExecuting"
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"id": "d756316f-e9e8-45f3-bbea-9b2fe6986f5a",
		"action": "WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)",
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7361828+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7486952+01:00",
	"log.level": "Information",
	"message": "Executed action \"WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)\" in 13.0702ms",
	"_metadata": {
		"message_template": "Executed action {ActionName} in {ElapsedMilliseconds}ms",
		"elapsed_milliseconds": 13.0702,
		"event_id": {
			"id": 2,
			"name": "ActionExecuted"
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"id": "d756316f-e9e8-45f3-bbea-9b2fe6986f5a",
		"action": "WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)",
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7486952+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7585408+01:00",
	"log.level": "Information",
	"message": "Executed endpoint '\"WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)\"'",
	"_metadata": {
		"message_template": "Executed endpoint '{EndpointName}'",
		"endpoint_name": "WebApplication3.Controllers.WeatherForecastController.Get (WebApplication3)",
		"event_id": {
			"id": 1,
			"name": "ExecutedEndpoint"
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7585408+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Routing.EndpointMiddleware",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}
{
	"@timestamp": "2020-03-11T12:31:42.7686387+01:00",
	"log.level": "Information",
	"message": "Request finished in 68.629ms 200 application/json; charset=utf-8",
	"_metadata": {
		"message_template": "{HostingRequestFinishedLog:l}",
		"elapsed_milliseconds": 68.629,
		"status_code": 200,
		"content_type": "application/json; charset=utf-8",
		"hosting_request_finished_log": "Request finished in 68.629ms 200 application/json; charset=utf-8",
		"event_id": {
			"id": 2
		},
		"request_id": "0HLU5LUMBRSDR:00000003",
		"request_path": "/weatherforecast",
		"span_id": "|a1999d5b-40d6b8dab8d21e2c.",
		"trace_id": "a1999d5b-40d6b8dab8d21e2c",
		"parent_id": "",
		"connection_id": "0HLU5LUMBRSDR"
	},
	"ecs": {
		"version": "1.4.0"
	},
	"event": {
		"severity": 2,
		"timezone": "W. Europe Standard Time",
		"created": "2020-03-11T12:31:42.7686387+01:00"
	},
	"log": {
		"logger": "Microsoft.AspNetCore.Hosting.Diagnostics",
		"original": null
	},
	"process": {
		"thread": {
			"id": 13
		},
		"pid": 15588,
		"name": "WebApplication3",
		"executable": "WebApplication3",
		"title": "WebApplication3"
	}
}

```
So, I'm confused as to how this is supposed to work. One possible issue I see is that my .csproj file has <TargetFramework>netcoreapp3.1</TargetFramework> so the conditional logic for getting HttpContext/HttpContextAccessor using #if NETSTANDARD does not kick in, but even when I rewrite the code to only support ".NET Core mode" it is still not able to get a reference to IHttpContextAccessor.

I have custom Serilog enrichers which use IHttpContextAccessor to get data from HttpContext, and they work fine, but I guess they use DI which does not seem to work for EcsTextFormatterConfiguration. Am I supposed to manually get IHttpContextAccessor from my DI container, and then pass it to MapHttpContext on the configuration objext? If so, then I guess configuring this with appsettings.json is out of the question? Maybe I'm just not understanding how this is supposed to work...

**Comment by codebrain** [comment]
Hi @olavaukan 

Thanks for raising, and with such clear reproduction steps. I have a reprodction working locally, and I believe I have the fix aswell.

I will update in here shortly.

**Comment by codebrain** [comment]
Hi @olavaukan - please see the PR here: https://github.com/elastic/ecs-dotnet/pull/63

Whilst the PR is open, you should be able to pull the respective branch (serilog-registration) and run the sample locally.

There are a couple of key points in the code to note here.

The creation of the `EcsTextFormatter` in `Program.cs`:
https://github.com/elastic/ecs-dotnet/blob/4a845a3115df4a1f407b0832b59548428816f700/examples/aspnetcore-with-serilog/Program.cs#L20-L40

The availability of the `HttpContextAccessor` in `Startup.cs`:

https://github.com/elastic/ecs-dotnet/blob/4a845a3115df4a1f407b0832b59548428816f700/examples/aspnetcore-with-serilog/Startup.cs#L19-L25

With those two things in place you should be all set and see the following (or close to) in the console:

![image](https://user-images.githubusercontent.com/148974/76587014-ac34c400-6536-11ea-9e55-c062447a7f6f.png)

...and the JSON containing the relevant Http information:

```
{
  "@timestamp": "2020-03-13T14:17:18.4976763+11:00",
  "log.level": "Information",
  "message": "Request finished in 353.60360000000003ms 200 application/json; charset=utf-8",
  "_metadata": {
    "message_template": "{HostingRequestFinishedLog:l}",
    "elapsed_milliseconds": 353.60360000000003,
    "status_code": 200,
    "content_type": "application/json; charset=utf-8",
    "hosting_request_finished_log": "Request finished in 353.60360000000003ms 200 application/json; charset=utf-8",
    "event_id": {
      "id": 2
    },
    "request_id": "0HLU6VK4GT6MU:00000001",
    "request_path": "/weatherforecast",
    "span_id": "|b4fce741-48d10adaba0a6f30.",
    "trace_id": "b4fce741-48d10adaba0a6f30",
    "parent_id": ""
  },
  "client": {
    "address": "::1",
    "ip": "::1"
  },
  "ecs": {
    "version": "1.5.0"
  },
  "event": {
    "severity": 2,
    "timezone": "AUS Eastern Standard Time",
    "created": "2020-03-13T14:17:18.4976763+11:00"
  },
  "http": {
    "request": {
      "method": "GET"
    },
    "response": {
      "status_code": 200
    }
  },
  "log": {
    "logger": "Microsoft.AspNetCore.Hosting.Diagnostics",
    "original": null
  },
  "process": {
    "thread": {
      "id": 10
    },
    "pid": 16656,
    "name": "AspnetCoreExample",
    "executable": "AspnetCoreExample",
    "title": ""
  },
  "server": {
    "domain": "localhost:5001"
  },
  "url": {
    "original": "/weatherforecast",
    "full": "https://localhost:5001/weatherforecast",
    "scheme": "https",
    "domain": "localhost:5001",
    "port": 5001,
    "path": "/weatherforecast",
    "query": ""
  },
  "user_agent": {
    "os": {
      "full": "Windows 10",
      "family": "Windows",
      "version": "10"
    },
    "device": {
      "name": "Other"
    },
    "original": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36",
    "name": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36",
    "version": "80.0.3987"
  }
}
```

Would you be able to review and see if this works for you - please drop any notes on the PR.

If we are good, then this can be merged and a patch release `1.4.3` can be created and pushed.

**Comment by codebrain** [comment]
I've also added an `/examples` folder as I believe it'll be good to demonstrate how all of these packages can come together: https://github.com/elastic/ecs-dotnet/tree/serilog-registration/examples/aspnetcore-with-serilog

**Comment by olavaukan** [comment]
Can confirm that this is fixed by #63

**Comment by codebrain** [comment]
Will be released in `1.4.3` packages. Thanks for your help @olavaukan
